design/00292: EEPROM storage.md

```diff
@@ -0,0 +1,64 @@
+* automatically save the hardware state of the machine
+* easily retrieve the hardware configuration for debugging
+* feed the metadata with precise hardware information
+
+## Background
+
+A specific HAT was developped to satisfy the PlanktoScope needs. 
+This PlanktoScope HAT connected to the raspberry pi and all of the external devices (led, motors, pump, camera) contains a 32 Kbit EEPROM chip which is a "type of non-volatile ROM that enables individual bytes of data to be erased and reprogrammed" (Wikipedia).
+It is therefore possible to store some data inside this chip and to read it back.
+
+For now (Sunray version 2024.0.0-beta.2) some of the hardware configuration of the PlanktoScope is entered by the user in the GUI.
+We would like to have an hardware configuration specific to each machine in case some specific modification is made.
+We would like to store this configuration in a safe place, independent from the software installed on the raspberry pi.
+
+## Proposal
+
+We therefore propose to write a python script that allows the storage and the reading of the PlanktoScope hardware information on the EEPROM chip.
+
+This information stored should contain : 
+* PlanktoScope reference
+* Plantoscope serial number
+* PlanktoScope version
+* Plantoscope fabrication date
+* HAT serial number
+* HAT version
+* Driver reference
+* Pump reference
+* Stepper focus reference
+* Objective lens reference
+* Tube lens reference
+* Flowcell thickness
+* LED reference
+
+Those informations could be used to feed the metadata stored in the hardware.json file as described in the issue 290 (https://github.com/PlanktoScope/PlanktoScope/issues/290).
+
+## Rationale
+
+The EEPROM chip currently used for the HAT has a capacity of 32Kbits. This provides adequate space for the necessary configuration without redundancy or excessive overhead.
+The use of our HAT is relevant as the PlanktoScope HAT is linked to every device of the machine. This allows unique data storage for each PlanktoScope unit, helping track individual hardware changes or updates.
+The EEPROM on the HAT is accessible directly through the I2C bus of the Raspberry Pi, simplifying software development.
+
+This solution is exclusive to the PlanktoScope HAT, limiting compatibility with other configurations (e.g., Adafruit HATs without EEPROM).
+Though currently sufficient, 32Kbits could become a constraint if future features or expanded data storage needs arise.
+
+## Compatibility
+
+The main compatibility issue concerns the adafruit HAT users.
+For those, a simple solution would be to continue selecting their configuration in the GUI, or adapt the script to store informations in the Raspberry pi EEPROM chip in case of not finding the PlanktoScope HAT EEPROM adress.
+The PlanktoScope versions V2.5 and V2.6 do not natively support writing data to the EEPROM chip because the write-protect (WP) pin of the EEPROM is enabled. To disable write protection, a solder point must be applied to the WP pin.
+
```

Comment:
This design document also needs to talk about data schema compatibility: if/when the metadata we store on the EEPROM changes (e.g. by adding or removing fields, or by adding new enum values for strings), how will we ensure backwards compatibility and/or forwards compatibility? I believe a common approach is to store a data schema version number as part of the data, as a way to detect when we might need to handle certain backwards/forwards-compatibility situations in certain ways.

Since there will be an MQTT API associated with this functionality, I also recommend addressing the question of API compatibility. For example, maybe this design document say the following things in the compatibility section:
1. This proposal will not modify any existing MQTT API routes, but it will add some new MQTT API routes related to EEPROM functionality.
2. All MQTT API routes added by this proposal are to be considered experimental, with no guarantees of compatibility in the future.